File: go/agents.go
package agenttrust

import (
	"context"
	"fmt"
	"net/http"
)

// AgentsAPI provides agent registration and lifecycle management.
//
// Agents are the core identity primitive in ATI. The platform does not issue
// certificates; use TokensAPI.Issue to mint opaque "at_" tokens for an agent.
type AgentsAPI struct {
	client *Client
}

// Create registers a new agent with the AgentTrust ID identity service.
//
// By default the SDK generates an Ed25519 identity keypair on this machine and
// registers only the public key; the returned Agent.PrivateKey holds the private
// key, which never leaves this process — persist it in a KeyStore (e.g.
// NewKeychainKeyStore). To bring your own key, set CreateAgentRequest.PublicKey,
// in which case the SDK sends only that and Agent.PrivateKey is empty. The
// platform does not issue certificates.
//
// Example:
//
//	agent, err := client.Agents.Create(ctx, ati.CreateAgentRequest{
//	    Name:         "my-assistant",
//	    Framework:    "langchain",
//	    Capabilities: []string{"files:read", "web:fetch"},
//	})
//	if err != nil {
//	    log.Fatal(err)
//	}
func (a *AgentsAPI) Create(ctx context.Context, req CreateAgentRequest) (*Agent, error) {
	framework := req.Framework
	if framework == "" {
		framework = "custom"
	}
	caps := req.Capabilities
	if caps == nil {
		caps = []string{}
	}
	meta := req.Metadata
	if meta == nil {
		meta = map[string]interface{}{}
	}

	// Generate an identity keypair client-side when the caller didn't supply a
	// public key, so the private key never leaves this process. Only the public
	// key is registered with the platform.
	publicKeyPEM := req.PublicKey
	var generatedPrivateKeyPEM string
	if publicKeyPEM == "" {
		kp, err := GenerateAgentKey()
		if err != nil {
			return nil, err
		}
		publicKeyPEM = kp.PublicKeyPEM
		generatedPrivateKeyPEM = kp.PrivateKeyPEM
	}

	apiReq := createAgentAPIRequest{
		Name:         req.Name,
		Framework:    framework,
		Capabilities: caps,
		Metadata:     meta,
		OrgID:        req.OrgID,
		PublicKey:    publicKeyPEM,
	}

	var resp createAgentResponse
	if err := a.client.doRequest(ctx, http.MethodPost, "/api/v1/agents", apiReq, &resp); err != nil {
		return nil, err
	}

	// The API may return the agent data in a nested "agent" field or at the top level.
	var agent *Agent
	if resp.Agent != nil {
		agent = parseAgentFromData(resp.Agent)
	} else {
		agent = parseAgentFromData(&agentData{
			ID:           resp.ID,
			Name:         resp.Name,
			OrgID:        resp.OrgID,
			Framework:    resp.Framework,
			PublicKey:    resp.PublicKey,
			Status:       resp.Status,
			Capabilities: resp.Capabilities,
			Metadata:     resp.Metadata,
			CreatedAt:    resp.CreatedAt,
			PrivateKey:   resp.PrivateKey,
		})
	}

	// When we generated the keypair locally, the private key stays here and is
	// never round-tripped through the platform.
	if generatedPrivateKeyPEM != "" && agent != nil {
		agent.PrivateKey = generatedPrivateKeyPEM
	}
	return agent, nil
}

// Get retrieves an agent by its ID.
//
// Returns a NotFoundError if the agent does not exist.
func (a *AgentsAPI) Get(ctx context.Context, agentID string) (*Agent, error) {
	var resp agentData
	path := fmt.Sprintf("/api/v1/agents/%s", agentID)
	if err := a.client.doRequest(ctx, http.MethodGet, path, nil, &resp); err != nil {
		return nil, err
	}
	return parseAgentFromData(&resp), nil
}

// List retrieves all agents, optionally filtered by organization ID.
//
// Pass an empty string for orgID to list all agents.
func (a *AgentsAPI) List(ctx context.Context, orgID string) ([]*Agent, error) {
	path := "/api/v1/agents"
	if orgID != "" {
		path += "?org_id=" + orgID
	}

	var resp agentListResponse
	if err := a.client.doRequest(ctx, http.MethodGet, path, nil, &resp); err != nil {
		return nil, err
	}

	agents := make([]*Agent, 0, len(resp.Agents))
	for i := range resp.Agents {
		agents = append(agents, parseAgentFromData(&resp.Agents[i]))
	}
	return agents, nil
}

// Revoke permanently revokes an agent and all its tokens.
//
// This action is immediate and cannot be undone.
func (a *AgentsAPI) Revoke(ctx context.Context, agentID string, reason string) error {
	if reason == "" {
		reason = "manual_revocation"
	}
	path := fmt.Sprintf("/api/v1/agents/%s/revoke", agentID)
	return a.client.doRequest(ctx, http.MethodPost, path, revokeRequest{Reason: reason}, nil)
}

File: rust/src/agents.rs
//! Agent registration and lifecycle management.
//!
//! Agents are the core identity primitive in ATI. Each agent gets an Ed25519
//! registration data upon creation.
//!
//! # Example
//!
//! ```rust,no_run
//! use agenttrustid::{AgentTrustClient, CreateAgentRequest};
//!
//! let client = AgentTrustClient::builder()
//!     .base_url("http://localhost:8080")
//!     .api_key("sk_live_xxx")
//!     .build()
//!     .unwrap();
//!
//! let agent = client.agents().create(&CreateAgentRequest {
//!     name: "my-assistant".to_string(),
//!     framework: "langchain".to_string(),
//!     capabilities: vec!["files:read".to_string(), "web:fetch".to_string()],
//!     ..Default::default()
//! }).unwrap();
//!
//! // Save agent.private_key securely!
//! println!("Agent ID: {}", agent.id);
//! ```

use crate::client::AgentTrustClient;
use crate::error::Result;
use crate::models::{
    Agent, AgentListResponse, CreateAgentRequest, CreateAgentResponse, RevokeAgentRequest,
};

/// Provides agent registration and lifecycle management.
///
/// Obtained via [`AgentTrustClient::agents()`].
pub struct AgentsAPI<'a> {
    pub(crate) client: &'a AgentTrustClient,
}

impl<'a> AgentsAPI<'a> {
    /// Register a new agent with the AgentTrust ID identity service.
    ///
    /// The returned [`Agent`] includes a `private_key` that is **only available
    /// during creation**. Store it securely; it cannot be retrieved again.
    ///
    /// # Errors
    ///
    /// - [`AgentTrustError::Validation`](crate::AgentTrustError::Validation) if required fields are missing.
    /// - [`AgentTrustError::Authentication`](crate::AgentTrustError::Authentication) if the API key is invalid.
    pub fn create(&self, req: &CreateAgentRequest) -> Result<Agent> {
        let resp: CreateAgentResponse = self.client.request("POST", "/api/v1/agents", Some(req))?;
        Ok(resp.into_agent())
    }

    /// Retrieve an agent by its unique ID.
    ///
    /// # Errors
    ///
    /// - [`AgentTrustError::NotFound`](crate::AgentTrustError::NotFound) if the agent does not exist.
    pub fn get(&self, agent_id: &str) -> Result<Agent> {
        let path = format!("/api/v1/agents/{}", agent_id);
        let resp: Agent = self.client.request("GET", &path, None::<&()>)?;
        Ok(resp)
    }

    /// List all agents, optionally filtered by organization ID.
    ///
    /// Pass `None` for `org_id` to list agents across all organizations.
    pub fn list(&self, org_id: Option<&str>) -> Result<Vec<Agent>> {
        let path = match org_id {
            Some(id) => format!("/api/v1/agents?org_id={}", id),
            None => "/api/v1/agents".to_string(),
        };
        let resp: AgentListResponse = self.client.request("GET", &path, None::<&()>)?;
        let agents = resp.agents.into_iter().map(|ad| ad.into_agent()).collect();
        Ok(agents)
    }

    /// Permanently revoke an agent and all its tokens.
    ///
    /// This action is immediate and cannot be undone.
    ///
    /// # Arguments
    ///
    /// * `agent_id` - The agent to revoke.
    /// * `reason` - Human-readable reason for the revocation.
    pub fn revoke(&self, agent_id: &str, reason: &str) -> Result<()> {
        let path = format!("/api/v1/agents/{}/revoke", agent_id);
        let body = RevokeAgentRequest {
            reason: reason.to_string(),
        };
        self.client.request_no_response("POST", &path, Some(&body))
    }
}

// Implement Default for CreateAgentRequest so users can use ..Default::default()
impl Default for CreateAgentRequest {
    fn default() -> Self {
        Self {
            name: String::new(),
            framework: "custom".to_string(),
            capabilities: Vec::new(),
            metadata: serde_json::Value::Object(serde_json::Map::new()),
            org_id: None,
        }
    }
}
